### PR TITLE
Fix Nullable Value Types.

### PR DESCRIPTION
**Pull request by mikernet**
Depends on PR https://github.com/dotnet/docfx/pull/7043 so merge that first.

Fixes https://github.com/dotnet/docfx/issues/7045
Fixes https://github.com/dotnet/docfx/issues/5964

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/docfx/pull/7054)

### CODE REVIEW

**Comment by mikernet** [comment]
@928PJY Can you review this PR and the one it depends on? Would really like to include this into the 3.0 release to fix these long standing issues. I can this version again all my libs that use docfx and the documentation is much improved.

**Comment by yufeih** [comment]
@mikernet , would you help resolve the conflicts in this PR now that #7043 is merged?

**Comment by mikernet** [comment]
> @mikernet , would you help resolve the conflicts in this PR now that #7043 is merged?

Yeah for sure, I'll give it a go here shortly.

**Comment by mikernet** [comment]
@yufeih I've confused myself as to what the best way to proceed here is. Should I merge dev onto this branch or rebase it somehow? Not sure how to do the latter since this branch was going off of the other one, and when I do interactive rebase onto dev it lists all the commits from the previous branch as well.

**Comment by yufeih** [comment]
> @yufeih I've confused myself as to what the best way to proceed here is. Should I merge dev onto this branch or rebase it somehow? Not sure how to do the latter since this branch was going off of the other one, and when I do interactive rebase onto dev it lists all the commits from the previous branch as well.

#7043 was squash merged into `dev`, so it becomes a single commit. Try cherry pick the two relevant commits, suppose you are on `fix-nvts`:

```bash
git reset --hard dev  # Hard reset to dev
git cherry-pick 96621be 6aad424 # Cherry-pick commits for nullable value type fix
git push -f # force push to sync the change
```

File: src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/CSYamlModelGenerator.cs [hunk 1]
```diff
@@ -870,16 +870,22 @@ private static ExpressionSyntax GetLiteralExpression(TypedConstant constant)
 
         private static ExpressionSyntax GetLiteralExpression(object value, ITypeSymbol type)
         {
+            bool isNullable = type.GetDocumentationCommentId().StartsWith("T:System.Nullable{");
```
Comment (yufeih) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/CSYamlModelGenerator.cs:873]: Is there a better way of detecting nullable value types? Using docId feels like a hack. `type.OriginalDefinition` seems to be always `System.Nullable<T>`?

Comment (mikernet) [reply src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/CSYamlModelGenerator.cs:873]: Not sure - I went with the technique that was currenty being used everywhere else to for type tests:

https://github.com/mikernet/docfx/blob/5150c5f87ee7bbf9922344dd4edb910a1aea8bee/src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/CSYamlModelGenerator.cs#L898

https://github.com/mikernet/docfx/blob/5150c5f87ee7bbf9922344dd4edb910a1aea8bee/src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/CSYamlModelGenerator.cs#L1136

https://github.com/mikernet/docfx/blob/5150c5f87ee7bbf9922344dd4edb910a1aea8bee/src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/CSYamlModelGenerator.cs#L1157

Etc...

Comment (yufeih) [reply src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/CSYamlModelGenerator.cs:873]: Then I'm good if this is an established way.

**Review by yufeih — APPROVED** [review]
Nice work 👍

**Comment by superyyrrzz** [comment]
@mikernet We reverted this change, as we find it will break building our own document site.

I added more checks in PR validation in #7285 to help identify such issues earlier. Could you take a look? A new PR is welcomed 😄 Thanks!

**Comment by mikernet** [comment]
What broke? I've been using a build that included my PRs for my own code bases and didn't have any problems so I'm curious what issue you ran into.

**Comment by superyyrrzz** [comment]
@mikernet Could you try build docfx's documentation? `docfx.exe Documentation/docfx.json`. This will meet some error like below:

```
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSYamlModelGenerator.GetLiteralExpression(Object value, ITypeSymbol type) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\CSYamlModelGenerator.cs:line 0
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSYamlModelGenerator.GetDefaultValueClause(IParameterSymbol symbol) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\CSYamlModelGenerator.cs:line 786
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSYamlModelGenerator.GetParameter(IParameterSymbol p, IFilterVisitor filterVisitor, Boolean isThisParameter) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\CSYamlModelGenerator.cs:line 746
   at System.Linq.Enumerable.<SelectIterator>d__5`2.MoveNext()
   at Microsoft.CodeAnalysis.CSharp.SyntaxFactory.SeparatedList[TNode](IEnumerable`1 nodes)
   at Microsoft.DocAsCode.Metadata.ManagedReference.CSYamlModelGenerator.GetMethodSyntax(IMethodSymbol symbol, IFilterVisitor filterVisitor) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\CSYamlModelGenerator.cs:line 392
   at Microsoft.DocAsCode.Metadata.ManagedReference.CompositeYamlModelGenerator.GenerateSyntax(MemberType type, ISymbol symbol, SyntaxDetail syntax, SymbolVisitorAdapter adapter) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\CompositeYamlModelGenerator.cs:line 96
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitMethod(IMethodSymbol symbol) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\SymbolVisitorAdapter.cs:line 272
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamedType(INamedTypeSymbol symbol) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\SymbolVisitorAdapter.cs:line 222
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\SymbolVisitorAdapter.cs:line 557
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitNamespace(INamespaceSymbol symbol) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\SymbolVisitorAdapter.cs:line 163
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitDescendants[T](IEnumerable`1 children, Func`2 getChildren, Func`2 filter) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\SymbolVisitorAdapter.cs:line 557
   at Microsoft.DocAsCode.Metadata.ManagedReference.SymbolVisitorAdapter.VisitAssembly(IAssemblySymbol symbol) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\Visitors\SymbolVisitorAdapter.cs:line 147
   at Microsoft.DocAsCode.Metadata.ManagedReference.RoslynMetadataExtractor.Extract(ExtractMetadataOptions options) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn\ExtractMetadata\RoslynMetadataExtractor.cs:line 47
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.GetMetadataFromProjectLevelCache(IBuildController controller, IInputParameters key) in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference\ExtractMetadataWorker.cs:line 485
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<SaveAllMembersFromCacheAsync>d__13.MoveNext() in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference\ExtractMetadataWorker.cs:line 300
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DocAsCode.Metadata.ManagedReference.ExtractMetadataWorker.<ExtractMetadataAsync>d__11.MoveNext() in D:\a\1\s\src\Microsoft.DocAsCode.Metadata.ManagedReference\ExtractMetadataWorker.cs:line 109
	72 Warning(s)
	1 Error(s)

```

**Comment by mikernet** [comment]
What is the story with the v3 branch? Is it going to inherit the work that is going into the dev branch or is it a completely separate thing?

**Comment by mikernet** [comment]
PR with fixes submitted here: https://github.com/dotnet/docfx/pull/7343

**Pull request merged**